### PR TITLE
fix for read_html with bs4 failing on table with header and one column

### PR DESCRIPTION
**Pull request by hnykda**
- [x] closes [#9178](https://github.com/pydata/pandas/issues/9178)
- [x] The test is added and passing (while failing before the fix).
- [x] passes `git diff upstream/master | flake8 --diff`
- [x] whatsnew entry

Fix as had been proposed in [PR 9194](https://github.com/pydata/pandas/pull/9194#issuecomment-209849563), but this PR was closed because of tests missing. They are added now.

### CODE REVIEW

**Comment by jreback** [comment]
``git diff master | flake8 --diff``

**Comment by hnykda** [comment]
Should be OK now.

File: pandas/io/tests/test_html.py [hunk 1]
```diff
@@ -416,6 +416,30 @@ def test_empty_tables(self):
         res2 = self.read_html(StringIO(data2))
         assert_framelist_equal(res1, res2)
 
+    def test_header_and_one_column(self):
+        """
+        Don't fail with bs4 when there is a header and only one column
+        """
+        data = StringIO('''<html>
```
Comment (jreback) [inline pandas/io/tests/test_html.py]: add the issue number as a comment

File: pandas/io/tests/test_html.py [hunk 2]
```diff
@@ -416,6 +416,30 @@ def test_empty_tables(self):
...
+                <tbody>
+                    <tr>
+                        <td>first</td>
+                    </tr>
+                </tbody>
+            </table>
+            </body>
+        </html>''')
+
+        correct_df = DataFrame(data={'Header': 'first'}, index=[0])
```
Comment (jreback) [inline pandas/io/tests/test_html.py]:
```
result = ....
expected = ...
tm.assert_frame_equal(result, expected
````

File: doc/source/whatsnew/v0.18.1.txt [hunk 3]
```diff
@@ -344,3 +344,5 @@ Bug Fixes
 
 
 - Bug in ``fill_value`` is ignored if the argument to a binary operator is a constant (:issue `12723`)
+
+- Bug in ``read_html`` when using bs4 flavor and parsing table with a header and only one column (:issue `9178`)
```
Comment (jreback) [inline doc/source/whatsnew/v0.18.1.txt]: ``pd..read_html()``

**Comment by jreback** [comment]
small comments. ping when green.

**Comment by hnykda** [comment]
Done. 

(I wasn't sure if I can use `(:issue 9178)`, so it's just as a regular comment)

File: pandas/io/tests/test_html.py [hunk 4]
```diff
@@ -416,6 +416,31 @@ def test_empty_tables(self):
...
 
+    def test_header_and_one_column(self):
+        """
+        Don't fail with bs4 when there is a header and only one column
+        as described in issue #9178
+        """
+        data = StringIO('''<html>
+            <body>
+             <table>
+                <thead>
```
Comment (jreback) [inline pandas/io/tests/test_html.py]: this doesn't seem to replicate the error message though:

```
In [1]:         data = StringIO('''<html>
   ...:             <body>
   ...:              <table>
   ...:                 <thead>
   ...:                     <tr>
   ...:                         <th>Header</th>
   ...:                     </tr>
   ...:                 </thead>
   ...:                 <tbody>
   ...:                     <tr>
   ...:                         <td>first</td>
   ...:                     </tr>
   ...:                 </tbody>
   ...:             </table>
   ...:             </body>
   ...:         </html>''')

In [2]: pd.read_html(data)
Out[2]: 
[  Header
 0  first]

In [3]: pd.read_html(data)[0]
Out[3]: 
  Header
0  first

In [4]: pd.__version__
Out[4]: u'0.18.0'
```

Comment (jreback) [inline pandas/io/tests/test_html.py]: should't this raise a similar error?

**Comment by hnykda** [comment]
Well, when I do:

```
In [2]: pandas.__version__
Out[2]: '0.18.0'

In [3]: s = '''<html>
            <body>
             <table>
                <thead>
                    <tr>
                        <th>Header</th>
                    </tr>
                </thead>
                <tbody>
                    <tr>
                        <td>first</td>
                    </tr>
                </tbody>
            </table>
            </body>
        </html>'''

In [4]: pandas.read_html(s, flavor="bs4")
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-4-5f8f3ea79c02> in <module>()
----> 1 pandas.read_html(s, flavor="bs4")

/home/dan/.local/opt/miniconda3/envs/mathbs/lib/python3.5/site-packages/pandas/io/html.py in read_html(io, match, flavor, header, index_col, skiprows, attrs, parse_dates, tupleize_cols, thousands, encoding)
    868     _validate_header_arg(header)
    869     return _parse(flavor, io, match, header, index_col, skiprows,
--> 870                   parse_dates, tupleize_cols, thousands, attrs, encoding)

/home/dan/.local/opt/miniconda3/envs/mathbs/lib/python3.5/site-packages/pandas/io/html.py in _parse(flavor, io, match, header, index_col, skiprows, parse_dates, tupleize_cols, thousands, attrs, encoding)
    741                                       parse_dates=parse_dates,
    742                                       tupleize_cols=tupleize_cols,
--> 743                                       thousands=thousands))
    744         except StopIteration:  # empty table
    745             continue

/home/dan/.local/opt/miniconda3/envs/mathbs/lib/python3.5/site-packages/pandas/io/html.py in _data_to_frame(data, header, index_col, skiprows, parse_dates, tupleize_cols, thousands)
    622 
    623     # fill out elements of body that are "ragged"
--> 624     _expand_elements(body)
    625 
    626     tp = TextParser(body, header=header, index_col=index_col,

/home/dan/.local/opt/miniconda3/envs/mathbs/lib/python3.5/site-packages/pandas/io/html.py in _expand_elements(body)
    599 
    600 def _expand_elements(body):
--> 601     lens = Series(lmap(len, body))
    602     lens_max = lens.max()
    603     not_max = lens[lens != lens_max]

/home/dan/.local/opt/miniconda3/envs/mathbs/lib/python3.5/site-packages/pandas/compat/__init__.py in lmap(*args, **kwargs)
    116 
    117     def lmap(*args, **kwargs):
--> 118         return list(map(*args, **kwargs))
    119 
    120     def lfilter(*args, **kwargs):

TypeError: len() of unsized object
```

while using patched version it works:
```
In [3]: import pandas

In [4]: pandas.__version__
Out[4]: '0.18.0+145.g9b6f9f2'

In [5]: s = '''<html>
            <body>
             <table>
                <thead>
                    <tr>
                        <th>Header</th>
                    </tr>
                </thead>
                <tbody>
                    <tr>
                        <td>first</td>
                    </tr>
                </tbody>
            </table>
            </body>
        </html>'''

In [6]: pandas.read_html(s, flavor="bs4")
Out[6]: 
[  Header
 0  first]
```

File: pandas/io/tests/test_html.py (hunk 4 above)
Comment (hnykda) [inline pandas/io/tests/test_html.py]: Well, when I do:

```
In [2]: pandas.__version__
Out[2]: '0.18.0'

In [3]: s = '''<html>
            <body>
             <table>
                <thead>
                    <tr>
                        <th>Header</th>
                    </tr>
                </thead>
                <tbody>
                    <tr>
                        <td>first</td>
                    </tr>
                </tbody>
            </table>
            </body>
        </html>'''

In [4]: pandas.read_html(s, flavor="bs4")
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-4-5f8f3ea79c02> in <module>()
----> 1 pandas.read_html(s, flavor="bs4")

/home/dan/.local/opt/miniconda3/envs/mathbs/lib/python3.5/site-packages/pandas/io/html.py in read_html(io, match, flavor, header, index_col, skiprows, attrs, parse_dates, tupleize_cols, thousands, encoding)
    868     _validate_header_arg(header)
    869     return _parse(flavor, io, match, header, index_col, skiprows,
--> 870                   parse_dates, tupleize_cols, thousands, attrs, encoding)

/home/dan/.local/opt/miniconda3/envs/mathbs/lib/python3.5/site-packages/pandas/io/html.py in _parse(flavor, io, match, header, index_col, skiprows, parse_dates, tupleize_cols, thousands, attrs, encoding)
    741                                       parse_dates=parse_dates,
    742                                       tupleize_cols=tupleize_cols,
--> 743                                       thousands=thousands))
    744         except StopIteration:  # empty table
    745             continue

/home/dan/.local/opt/miniconda3/envs/mathbs/lib/python3.5/site-packages/pandas/io/html.py in _data_to_frame(data, header, index_col, skiprows, parse_dates, tupleize_cols, thousands)
    622 
    623     # fill out elements of body that are "ragged"
--> 624     _expand_elements(body)
    625 
    626     tp = TextParser(body, header=header, index_col=index_col,

/home/dan/.local/opt/miniconda3/envs/mathbs/lib/python3.5/site-packages/pandas/io/html.py in _expand_elements(body)
    599 
    600 def _expand_elements(body):
--> 601     lens = Series(lmap(len, body))
    602     lens_max = lens.max()
    603     not_max = lens[lens != lens_max]

/home/dan/.local/opt/miniconda3/envs/mathbs/lib/python3.5/site-packages/pandas/compat/__init__.py in lmap(*args, **kwargs)
    116 
    117     def lmap(*args, **kwargs):
--> 118         return list(map(*args, **kwargs))
    119 
    120     def lfilter(*args, **kwargs):

TypeError: len() of unsized object
```

while using patched version it works:
```
In [3]: import pandas

In [4]: pandas.__version__
Out[4]: '0.18.0+145.g9b6f9f2'

In [5]: s = '''<html>
            <body>
             <table>
                <thead>
                    <tr>
                        <th>Header</th>
                    </tr>
                </thead>
                <tbody>
                    <tr>
                        <td>first</td>
                    </tr>
                </tbody>
            </table>
            </body>
        </html>'''

In [6]: pandas.read_html(s, flavor="bs4")
Out[6]: 
[  Header
 0  first]
```

**Comment by jreback** [comment]
@hnykda ahh I see. we are testing with multiple flavors. I think we default to ``lxml`` which I have installed so it works now. ok. then. ping on green.

**Comment by hnykda** [comment]
Exactly. 

Test passed :).

**Comment by jreback** [comment]
thanks @hnykda

**Pull request closed without merging**